src/Oscoin/Data/Core.hs

```diff
@@ -0,0 +1,502 @@
+
+deriving instance
+    ( Show (Signature c)
+    , Show (BlockHash c)
+    , Show (PublicKey c)
+    , Crypto.HasHashing c
+    ) => Show (Contribution c)
+
+deriving instance
+    ( Eq (Signature c)
+    , Eq (BlockHash c)
+    , Eq (PublicKey c)
+    ) => Eq (Contribution c)
+
+deriving instance
+    ( Ord (Signature c)
+    , Ord (BlockHash c)
+    , Ord (PublicKey c)
+    ) => Ord (Contribution c)
+
+-------------------------------------------------------------------------------
+
+-- | An update to a project dependency.
+data DependencyUpdate c =
+      Depend       (Address c) (Crypto.Hash c)
+    -- ^ Start depending on a project starting from the given state hash.
+    | Undepend     (Address c)
+    -- ^ Stop depending on a project.
+    deriving (Generic)
+
+deriving instance (Eq (Crypto.Hash c))   => Eq   (DependencyUpdate c)
+deriving instance (Ord (Crypto.Hash c))  => Ord  (DependencyUpdate c)
+deriving instance (Show (Crypto.Hash c)) => Show (DependencyUpdate c)
+
+-- | Additional signatures used to authorize a transaction in a
+-- *multi-sig* scenario.
+newtype Signatures c = Signatures [Signed () c]
+
+-------------------------------------------------------------------------------
+-- Project
+-------------------------------------------------------------------------------
+
+-- | A project.
+data Project c = Project
+    { pAccount         :: Account c
+    -- ^ The project account or *fund*.
+
+    , pMaintainers     :: Map (Address c) (Member c)
+    , pContributors    :: Map (Address c) (Member c)
+    , pSupporters      :: Map (Address c) (Member c)
+    , pDependencies    :: Map (Address c) (Dependency c)
+
+    , pCheckpoints     :: Set (Checkpoint c)
+
+    -- * Contract
+    , pSendTransfer    :: SendTransfer' c
+    , pReceiveTransfer :: ReceiveTransfer' c
+    , pReceiveReward   :: ReceiveReward' c
+    , pCheckpoint      :: Checkpoint' c
+    , pUnregister      :: Unregister' c
+    , pAuthorize       :: Authorize' c
+    , pDeauthorize     :: Deauthorize' c
+    , pUpdateContract  :: UpdateContract' c
+    }
+
+defaultProject :: Ord (Crypto.Hash c) => Address c -> Project c
+defaultProject addr = Project
+    { pAccount          = mkAccount addr
+    , pMaintainers      = mempty
+    , pContributors     = mempty
+    , pSupporters       = mempty
+    , pDependencies     = mempty
+    , pCheckpoints      = mempty
+    , pSendTransfer     = defaultSendTransfer
+    , pReceiveTransfer  = defaultReceiveTransfer
+    , pReceiveReward    = defaultReceiveReward
+    , pCheckpoint       = defaultCheckpoint
+    , pUnregister       = defaultUnregister
+    , pAuthorize        = defaultAuthorize
+    , pDeauthorize      = defaultDeauthorize
+    , pUpdateContract   = defaultUpdateContract
+    }
+
+-- | Get the address of a project.
+projectAddr :: Project c -> Address c
+projectAddr = accountAddr . pAccount
+
+-- | A delegation of oscoin between two accounts.
+data Delegation c = Delegation
+    { dDelegator  :: Address c
```

Comment:
I fear that using one letter as the field prefix quickly lead to clashes if we introduce another data type starting with “D”.